### DISCUSSION TITLE
The proper way to use GDB via SSH from Windows to Linux host?

### DISCUSSION THREAD

**Discussion by Igorprostoff** [NationalSecurityAgency/ghidra#3238] [Debugger]
While trying to solve a simple Crackme, written for Linux, I wanted to try out the new debugger feature by using the "GDB via SSH" option in Ghidra, by connecting from Windows host (where the Ghidra is running) to the WSL2 host, where the GDB is able to debug the Crackme. But after I successfully connect from Ghidra to the Linux host with my login and password and provide the executable path as a command-line parameter, the only thing I see is a program output in the Interpreter window.
Is this the right way to use this feature or am I doing something wrong?

**Comment by d-millar** [comment]
@Igorprostoff The fact that you have output to the Interpreter window is a good sign - means you’re half way there.  Do you have a tree in the Objects window? If so, ids the tree opened up to the process.  (It’s supposed to do this automatically, but doesn’t always seem to work correctly.)  Opening to the process will trigger the recording, which in turn enables all the other windows.  (Yeah, I know - not the greatest design.  We’re working on fixing it.)

**Reply by Igorprostoff** [reply]
Thanks a lot for the quick response!
Yes, I can see the Objects window with one process in the "inferiors" section, but after trying to expand it's content (such as breakpoints, memory, and modules) I see an error "Request for children timed - out - try refreshing the node" and nothing more happens.

Also, I forgot to mention, that there is an error, occurring right after I connect to the host with SSH:
```
Command completed with <GdbCommandDoneEvent {} >, not any of [class agent.gdb.manager.evt.GdbCommandRunningEvent]
java.lang.IllegalStateException: Command completed with <GdbCommandDoneEvent {} >, not any of [class agent.gdb.manager.evt.GdbCommandRunningEvent]
	at agent.gdb.manager.impl.GdbPendingCommand.checkCompletion(GdbPendingCommand.java:234)
	at agent.gdb.manager.impl.cmd.AbstractLaunchGdbCommand.complete(AbstractLaunchGdbCommand.java:52)
	at agent.gdb.manager.impl.cmd.AbstractLaunchGdbCommand.complete(AbstractLaunchGdbCommand.java:22)
	at agent.gdb.manager.impl.GdbPendingCommand.finish(GdbPendingCommand.java:71)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1736)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)

---------------------------------------------------
Build Date: 2021-Jul-08 1512 EDT
Ghidra Version: 10.0.1
Java Home: C:\Program Files\AdoptOpenJDK\jdk-11.0.10.9-hotspot
JVM Version: AdoptOpenJDK 11.0.10
OS: Windows 10 10.0 amd64
Workstation: IGOR-PC.mshome.net
```

But this error doesn't always appear.

**Reply by d-millar** [reply]
Hmmm, well, that isn't as good.  :) 

A couple of more questions, I don't think this matters for the SSH case but version of gdb in WSL2?  Also, can you execute standard gdb commands in the Interpreter window or do you just see output?  Also, I guess, what is the output?

**Reply by Igorprostoff** [reply]
The GDB version is: GNU gdb (Ubuntu 9.2-0ubuntu1~20.04) 9.2
The execution of regular GDB commands in the interpreter fails, I get no output, here is the interpreter output:
![изображение](https://user-images.githubusercontent.com/33811302/126382099-1b27f1eb-32f7-4dc0-b60f-830f057f3876.png)

**Comment by d-millar** [comment]
Hmmm, do you have the same issues running something mundane, like plums or xclock?

**Reply by Igorprostoff** [reply]
If you mean running regular commands in the Interpreter, then yes, I still have no reaction, even no error messages:
![изображение](https://user-images.githubusercontent.com/33811302/126405140-5e6400fc-2f17-46db-9c42-dce8363e08c4.png)

I can't even quit GDB, just like I showed in the previous screenshot, the command "q" doesn't work.

**Reply by d-millar** [reply]
Sorry, wasn’t clear - I meant if you pick something like xclock or pluma as your initial target.  Just wondering if there was something specific to the crackme going on that was causing it to terminate.   A related question: debugging this with straight gdb raises no issues, I assume?

**Reply by Igorprostoff** [reply]
I copied the xclock binary from my WSL2 host, opened it in Ghidra and started debugging it via SSH, but I only got these outputs:
![изображение](https://user-images.githubusercontent.com/33811302/126515039-10ef0d1d-92c8-4432-a746-fe33e8a83fbc.png)

and 

![2021-07-21_18-15-50](https://user-images.githubusercontent.com/33811302/126515103-7d8f1f86-e67f-4d82-bbe4-272b8a289778.png)

which is more, than I had, while debugging my Crackme, but still, other windows (registers, threads, stack...) are empty.

Debugging the crackme with GDB works well, I was able to do that several times using CLI, while I was trying to figure out, what is the problem with Ghidra, working with GDB.

**Reply by d-millar** [reply]
OK, more stupid questions:  do the following commands work in your gdb?

show os
show version
show arch
show endian
info proc mappings
maintenance info sections

They all 'should' but maybe I'm going to be surprised.

Our best guess on the crackme vs xclock is the debugger is that the crackme has no main and, for reasons we don't quite understand, failed to start up using "starti", fell back to "start", and was actually running when you tried to enter the various commands.  A way to check this would be to get to the same point and use Ctrl-I (I think - ctrl-c, unfortunately is "copy" everywhere) to see if the process will break.  Alternatively, you could connect using SSH via the Targets window, "file" you executable in the Interpreter, and then "starti" - all of which is admittedly not very convenient.

This might explain the behavior you saw, but doesn't tell us why the recorder didn't start, which in turn enables all the other windows.  (More detail in the luma3ds Discussion on Github if you have time to kill.)

**Reply by Igorprostoff** [reply]
All of the mentioned commands work when I debug xclock, but they don't when I debug the crackme and I guess, that your conjecture about the problem with the "main" function is right, because, as far as I can understand, the crackme just handles after printing the banner because it is listening to the TCP ports (and it also reacts to connection to them, by printing more information), but gdb still can't execute any commands, which are sent to it with CLI. This happens also when I try to debug it with GDB without ghidra, by using the command "start" instead of "starti".
Also, the Ctrl+I combination doesn't help in this situation and it seems that Ghidra writes the line "(gdb)" before the commands, so the user thinks that he can execute the commands, while he can't.
Here is the attempt to execute commands with Ghidra interpreter:
![2021-07-21_23-40-04](https://user-images.githubusercontent.com/33811302/126557601-0efcc5f9-e5a1-4eec-8e99-713d0290b6c0.png)
And here is the attempts to execute commands with CLI of WSL2 :
![2021-07-21_23-36-55](https://user-images.githubusercontent.com/33811302/126557605-9031995e-9cdc-4666-ade9-4c4aa059e973.png)

The solution with manually setting the exec-file and starting its execution with the "starti" command from the ghidra interpreter works partly because I can see more information about the program in the "Objects" tab, but I can't see the dynamic listing of the program.
I also tried to debug this crackme on my WSL2 with Ghidra by connecting to the host with VNC and launching the GDB locally IN-VM and it works fine - I can see the dynamic listing, the stack info, threads, breakpoints, etc. And it works OK without manually starting the execution with the "starti" command like I have to do on Windows with an SSH connection to Linux.

**Reply by d-millar** [reply]
Wow, ok - thanks for trying all this.  All your information helps a lot.  Will try to reproduce these errors on WSL2 tomorrow if I can.  The VNC test was particularly interesting - good call. Am out of ideas for the moment, but will keep banging on this.  Would definitely like to get it solved.

**Reply by d-millar** [reply]
OK, we found the problem, and we definitely owe you a beer.  It's a major problem, and I'm somewhat surprised it hasn't affected a lot of other folks.  It will be fixed in 10.0.2.  That said, if you're building and/or running your own copy out of Eclipse, I can point you to the code to modify.  It's a one line fix, namely, in GdbManagerImpl::doExecute move the end brace after "curCmd = pcmd;" (around line 726?) to the end of the thenAccept block (around line 743?).  Points to @nsadeveloper789 for id'ing the bug and the fix!  After the fix, I tested it for the Launch case and the "file"+"starti" case, and both populated all the windows including the Dynamic listing.  You may still need to map the executable in Modules to get the Static listing to track the Dynamic listing because the paths will be different across the WSL/Windows boundary (may).

**Reply by Igorprostoff** [reply]
Wow, thank you for help with this issue! Glad to hear that my case helped to find such kind of problem😊.

**Reply by astrelsky** [reply]
> OK, we found the problem, and we definitely owe you a beer.  It's a major problem, and I'm somewhat surprised it hasn't affected a lot of other folks.  It will be fixed in 10.0.2.

Bystander effect.

**Comment by ryanmkurtz** [comment]
Fixed by 19fa121b30c4308680c5663e88b330ba119d79f7
